Integration tests

bokobza opened the pull request:
This change:
* Makes sure that all the tests go into a folder in TestCase/_class_name_/_test_name_.
* Moves a couple of non-integration tests to their rightful place
* now uses a bitcoind file that is included in a folder rather than downloading it from the internet. This will save time when the tests are run on a CI server.
* added some small improvements

This is a WIP, but this can already be reviewed and merged.

majikandy commented on src/Stratis.Bitcoin.IntegrationTests/BlockStore/BlockStoreTests.cs:
```diff
@@ -26,121 +27,69 @@ public BlockStoreTests()
         public void BlockRepositoryPutBatch()
         {
-            using (var dir = TestDirectory.Create())
+            using (var blockRepo = new BlockRepository(Network.Main, TestBase.CreateDataFolder(this), DateTimeProvider.Default, this.loggerFactory))
```
`blockRepo` -> `blockRepository`  ?

majikandy commented on src/Stratis.Bitcoin.IntegrationTests/BlockStore/BlockStoreTests.cs:
```diff
@@ -26,121 +27,69 @@ public BlockStoreTests()
-                using (var blockRepo = new BlockRepository(Network.Main, dir.FolderName, DateTimeProvider.Default, this.loggerFactory))
+                blockRepo.SetTxIndexAsync(true).Wait();
+
+                var lst = new List<Block>();
```
perhaps `lst` -> `blocks`

majikandy commented on src/Stratis.Bitcoin.IntegrationTests/BlockStore/BlockStoreTests.cs:
```diff
@@ -26,121 +27,69 @@ public BlockStoreTests()
+                for (int i = 0; i < 5; i++)
                 {
-                    blockRepo.SetTxIndexAsync(true).Wait();
+                    // put
```
is this in the right place ?  doesn't seem to tally with the line or the block of code

majikandy commented on src/Stratis.Bitcoin.IntegrationTests/BlockStore/BlockStoreTests.cs:
```diff
@@ -26,121 +27,69 @@ public BlockStoreTests()
+                blockRepo.PutAsync(lst.Last().GetHash(), lst).GetAwaiter().GetResult();
 
-                    blockRepo.PutAsync(lst.Last().GetHash(), lst).GetAwaiter().GetResult();
+                // check each block
```
check for what ?

majikandy commented:
It looks ok - couple of comments on naming. I haven't actually run the tests, just eyeballed the code

mikedennis commented on src/Stratis.Bitcoin.IntegrationTests/SegWitTests.cs:
```diff
@@ -14,9 +14,9 @@ public class SegWitTests
+            using (NodeBuilder builder = NodeBuilder.Create(this))
             {
-                CoreNode coreNode = builder.CreateBitcoinCoreNode();
+                CoreNode coreNode = builder.CreateBitcoinCoreNode(version: "0.15.1");
```
This appears to be the only test that uses the bitcoind? Should default version to "0.15.1" so that it's easy to updated all instances to a new version. Not really a problem right now as there is only this test, but maybe in future we will have more.

bokobza commented:
The last few changes are the following:
- Moved the RPC tests from NBitcoin.Tests to the Integration Tests project.
- Moved to the right class a few tests that were in the wrong class.
- Added NBitcoin.Tests to CI.
- Made the retrieval of test data files environment agnostic (Path.Combine("data","filename") instead of data\filename).
- Remove the Nodebuilder in NBitcoin.Tests.

Now, the only existing Nodebuilder is the one in Integration Tests.
All other tests are non-integration.

bokobza commented:
At the moment, the following tests are failing:
On AppVeyor
![image](https://user-images.githubusercontent.com/1867877/40422028-d77f7420-5e85-11e8-80e9-e3c856a075f6.png)

On Travis:
```
Failed   CanParseIpEndpoint
Error Message:  Assert.Throws() Failure
Expected: typeof(System.Net.Sockets.SocketException)
Actual:   typeof(System.Net.Internals.SocketExceptionFactory+ExtendedSocketException): Device not configured
```
+
the 3 tests requiring libbitcoinconsensus-0.dll to be present.

bokobza commented:
@dangershony can you please check this out, I need to move on from this.
Never mind the failing tests, I'll take care of them very soon.

bokobza commented on src/Stratis.Bitcoin.IntegrationTests/RPC/RPCTests.cs:
```diff
@@ -185,5 +211,130 @@ public void CanGenerateByStringArgs()
             Assert.StartsWith("[" + Environment.NewLine + "  \"", resp);
         }
+
+        // TODO: implement the RPC methods used below
```
yeah this is stuff that potentially needs to be done.

monsieurleberre commented on src/Stratis.Bitcoin.IntegrationTests/API/ApiSteps.cs:
```diff
@@ -40,7 +41,7 @@ public ApiSpecification(ITestOutputHelper output) : base(output)
         {
             this.httpClient = new HttpClient();
-            this.nodeGroupBuilder = new NodeGroupBuilder(this.CurrentTest.DisplayName);
+            this.nodeGroupBuilder = new NodeGroupBuilder(Path.Combine(this.GetType().Name, this.CurrentTest.DisplayName));
```
Actually there is a way to do it automatically : 
public class MyClass
{
  public void MyMethod([CallerFilePath]string callerFilePath = null, [CallerMemberName]string callerMemberName = null)
  {
    var callerTypeName = Path.GetFileNameWithoutExtension(callerFilePath);
    Console.WriteLine(callerTypeName);
    Console.WriteLine(callerMemberName);
  }
}

bokobza replied:
ah yes I had this at some point. I can't remember why I didn't use it, but let's revisit this as well.

bokobza commented on src/NBitcoin.Tests/pos_transaction_tests.cs:
```diff
@@ -3075,6 +3075,26 @@ public void test_IsStandard()
             Assert.True(!StandardScripts.IsStandardTransaction(t));
         }
 
+        //[Fact]
```
not yet, I want to reconsider that in the future.

bokobza commented on src/NBitcoin.Tests/TestUtils.cs:
```diff
@@ -2,6 +2,7 @@
 using System.IO;
 using System.Threading;
 using NBitcoin.DataEncoders;
+using Xunit;
```
fixed

bokobza commented on src/Stratis.Bitcoin.IntegrationTests.Common/TestHelper.cs:
```diff
@@ -3,9 +3,9 @@
 using System.Linq;
```
yeah definitely. In another PR though, this is becoming too big 🚴

bokobza commented:
Thanks for the comments @monsieurleberre, all good and valid.
Let's take care of this next time though, there is still a lot more that I'd like to change there.
